omnigibson/simulator.py

```diff
@@ -376,149 +367,78 @@ def remove_object(self, obj):
-        n_objects_to_initialize = len(self._objects_to_initialize)
-        if n_objects_to_initialize > 0 and self.is_playing():
-            # We iterate through the objects to initialize
-            # Note that we don't explicitly do for obj in self._objects_to_initialize because additional objects
-            # may be added mid-iteration!!
-            # For this same reason, after we finish the loop, we keep any objects that are yet to be initialized
-            for i in range(n_objects_to_initialize):
-                obj = self._objects_to_initialize[i]
-                obj.initialize()
-                if len(obj.states.keys() & self.object_state_types_on_contact) > 0:
-                    self._objects_require_contact_callback = True
-                if len(obj.states.keys() & self.object_state_types_on_joint_break) > 0:
-                    self._objects_require_joint_break_callback = True
-
-            self._objects_to_initialize = self._objects_to_initialize[n_objects_to_initialize:]
-
-        # Propagate states if the feature is enabled
-        if gm.ENABLE_OBJECT_STATES:
-            # Step the object states in global topological order (if the scene exists).
-            if self.scene is not None:
+        # If we don't have a valid scene, immediately return
+        if self._scene is None:
+            return
+
+        # Update omni
+        self._omni_update_step()
+
+        # If we're playing we, also run additional logic
+        if self.is_playing():
+            # Check to see if any objects should be initialized (only done IF we're playing)
+            n_objects_to_initialize = len(self._objects_to_initialize)
+            if n_objects_to_initialize > 0 and self.is_playing():
+                # We iterate through the objects to initialize
+                # Note that we don't explicitly do for obj in self._objects_to_initialize because additional objects
+                # may be added mid-iteration!!
+                # For this same reason, after we finish the loop, we keep any objects that are yet to be initialized
+                for i in range(n_objects_to_initialize):
+                    obj = self._objects_to_initialize[i]
+                    obj.initialize()
+                    if len(obj.states.keys() & self.object_state_types_on_contact) > 0:
+                        self._objects_require_contact_callback = True
+                    if len(obj.states.keys() & self.object_state_types_on_joint_break) > 0:
+                        self._objects_require_joint_break_callback = True
+
+                self._objects_to_initialize = self._objects_to_initialize[n_objects_to_initialize:]
+
+                # Re-initialize the physics view because the number of objects has changed
+                RigidContactAPI.initialize_view()
+
+                # Also refresh the transition rules that are currently active
+                TransitionRuleAPI.refresh_all_rules(objects=self.scene.objects)
```

Comment:
nit: should we package these two lines into a helper function?

Reply:
maybe something like `_sync_apis`?

Reply:
can't do utility function, the two lines aren't the same